### PR TITLE
test/e2e: more flexible pod spec generation

### PR DESCRIPTION
**Pull request by marquiz**
Change the pod spec generator functions to accept parameterization in
the form of more generic "mutator functions". This makes the addition of
new test specific pod spec customizations a lot cleaner. Plus, hopefully
makes the code a bit more readable as well.

Also, slightly simplify the SpecWithConfigMap() but dropping one
redundant argument.

Inspired by latest contributions by Talor Itzhak (titzhak@redhat.com).

### CODE REVIEW

**Comment by marquiz** [comment]
/assign @ArangoGutierrez @fmuyassarov 

ping @Tal-or

File: test/e2e/utils/pod.go [hunk 1]
```diff
@@ -141,27 +142,20 @@ func NFDMasterPod(image string, onMasterNode bool) *corev1.Pod {
 			RestartPolicy:      corev1.RestartPolicyNever,
 		},
 	}
-	if onMasterNode {
```
Comment (Tal-or) [inline test/e2e/utils/pod.go:144]: Was it removed on purpose?

**Review by Tal-or — COMMENTED** [review]
Very nice.
A small enhancement proposal from me about readability:
You can have a separate package named `pod` or `podspec` under `utils` and place `NFDWorkerPod` `NFDMasterPod`, etc. functions there.

Then you can rename the functions and call them from the tests in the following manner:
instead of:
```
testutils.NFDWorkerPod(opts...)
testutils.NFDMasterPod(opts...)
testutils.SpecWithContainerImage(...)
```

You'll have:
```
pod.NFDWorker(opts...)
pod.NFDMaster(opts...)
pod.SpecWithContainerImage(...)
```
It will keep the names shorter, clearer, and make the package more isolated and portable

File: test/e2e/utils/pod.go (hunk 1 above)
Comment (marquiz) [reply test/e2e/utils/pod.go:144]: Yeah, it's replaced by `SpecWithMasterNodeSelector`. It's currently unused so it could be dropped, too 🤔

**Comment by marquiz** [comment]
Thanks for the review @Tal-or

> A small enhancement proposal from me about readability:
> You can have a separate package named `pod` or `podspec` under `utils` and place `NFDWorkerPod` `NFDMasterPod`, etc. functions there.

Good note, I agree.

How about doing this refactoring in a separate PR? I think there are other things to be done on this front, e.g. separating the daemonset functionality in a separate package (or a separate file at least). Another thing that I spotted was that we could simplify the "sleeper pod" creation, basically set resource limits with a dedicated `PodSpecOption`. We could track these in a separate issue

**Comment by Tal-or** [comment]
> Thanks for the review @Tal-or
> 
> > A small enhancement proposal from me about readability:
> > You can have a separate package named `pod` or `podspec` under `utils` and place `NFDWorkerPod` `NFDMasterPod`, etc. functions there.
> 
> Good note, I agree.
> 
> How about doing this refactoring in a separate PR? I think there are other things to be done on this front, e.g. separating the daemonset functionality in a separate package (or a separate file at least). Another thing that I spotted was that we could simplify the "sleeper pod" creation, basically set resource limits with a dedicated `PodSpecOption`. We could track these in a separate issue

Sure. I can handle it if you want.

**Comment by marquiz** [comment]
> Sure. I can handle it if you want.

Yeah, sure go ahead. Would be appreciated 👍

**Comment by marquiz** [comment]
Updated the PR, converting nfd-worker tests to also use `SpecWithConfigMap()`

**Comment by Tal-or** [comment]
Sure, go ahead

**Comment by marquiz** [comment]
ping @ArangoGutierrez @fmuyassarov

**Review by ArangoGutierrez — APPROVED** [review]
/lgtm

**Pull request merged**